File: pipeline/reimpl/cocrdslc.py
"""
Reimplementation of COCRDSLC — CardDemo Credit Card View/Select Screen.

CICS online program. Displays full detail of a credit card by card number.
Also shows the associated account and customer information.

Navigation:
  ENTER → look up card by number
  PF3   → return to calling program
"""

from __future__ import annotations
from dataclasses import dataclass
from typing import Optional

from .carddemo_data import (
    CarddemoCommarea, ScreenHeader, CardRecord, AccountRecord,
    CustomerRecord, CardXrefRecord,
    CCDA_MSG_INVALID_KEY, DFHENTER, DFHPF3,
)


WS_PGMNAME = "COCRDSLC"
WS_TRANID  = "CCDL"


@dataclass
class CardDetailData:
    card_num: str = ""
    card_acct_id: int = 0
    card_cvv: int = 0
    card_embossed_name: str = ""
    card_expiration_date: str = ""
    card_active_status: str = ""
    acct_curr_bal: str = ""
    acct_credit_limit: str = ""
    cust_name: str = ""
    cust_id: int = 0


@dataclass
class CardViewResult:
    card_data: Optional[CardDetailData] = None
    commarea: Optional[CarddemoCommarea] = None
    screen: Optional[ScreenHeader] = None
    message: str = ""
    error: bool = False
    xctl_program: str = ""
    return_to_prev: bool = False


class CardRepository:
    def __init__(self, cards: dict[str, CardRecord]):
        self._cards = cards  # keyed by card_num

    def find(self, card_num: str) -> Optional[CardRecord]:
        return self._cards.get(card_num.strip())


class AccountRepository:
    def __init__(self, accounts: dict[int, AccountRecord]):
        self._accounts = accounts

    def find(self, acct_id: int) -> Optional[AccountRecord]:
        return self._accounts.get(acct_id)


class XrefRepository:
    def __init__(self, xrefs: dict[str, CardXrefRecord]):
        self._by_card = xrefs

    def find_by_card(self, card_num: str) -> Optional[CardXrefRecord]:
        return self._by_card.get(card_num.strip())


class CustomerRepository:
    def __init__(self, customers: dict[int, CustomerRecord]):
        self._customers = customers

    def find(self, cust_id: int) -> Optional[CustomerRecord]:
        return self._customers.get(cust_id)


def process_card_view(
    eibcalen: int,
    eibaid: str,
    commarea: CarddemoCommarea,
    card_num_input: str,
    card_repo: CardRepository,
    account_repo: AccountRepository,
    xref_repo: XrefRepository,
    customer_repo: CustomerRepository,
    preloaded_card_num: str = "",
) -> CardViewResult:
    """Process card view screen — mirrors COCRDSLC PROCEDURE DIVISION."""
    result = CardViewResult()

    if eibcalen == 0:
        result.return_to_prev = True
        result.xctl_program = "COSGN00C"
        return result

    result.commarea = commarea

    if eibaid in (DFHPF3, "PF3"):
        back = commarea.cdemo_from_program or "COMEN01C"
        result.xctl_program = back
        result.return_to_prev = True
        result.commarea = commarea
        return result

    if commarea.cdemo_pgm_context == 0:
        commarea.cdemo_pgm_context = 1
        search_num = preloaded_card_num or card_num_input
        if search_num.strip():
            _lookup_card(search_num, card_repo, account_repo, xref_repo, customer_repo, result)
        result.screen = ScreenHeader.now(WS_TRANID, WS_PGMNAME)
        result.commarea = commarea
        return result

    return _process_enter(
        card_num_input, card_repo, account_repo, xref_repo, customer_repo, commarea, result
    )


def _process_enter(card_num_input, card_repo, account_repo, xref_repo, customer_repo, commarea, result):
    if not card_num_input or card_num_input.strip() == "":
        result.error = True
        result.message = "Card number not provided"
        result.screen = ScreenHeader.now(WS_TRANID, WS_PGMNAME)
        result.screen.errmsg = result.message
        result.commarea = commarea
        return result

    _lookup_card(card_num_input, card_repo, account_repo, xref_repo, customer_repo, result)
    result.screen = ScreenHeader.now(WS_TRANID, WS_PGMNAME)
    result.screen.errmsg = result.message
    result.commarea = commarea
    return result


def _lookup_card(card_num, card_repo, account_repo, xref_repo, customer_repo, result):
    card = card_repo.find(card_num)
    if card is None:
        result.error = True
        result.message = f"Card {card_num.strip()} not found"
        return

    acct = account_repo.find(card.card_acct_id)
    xref = xref_repo.find_by_card(card.card_num)
    cust = None
    if xref:
        cust = customer_repo.find(xref.xref_cust_id)

    result.card_data = CardDetailData(
        card_num=card.card_num,
        card_acct_id=card.card_acct_id,
        card_cvv=card.card_cvv_cd,
        card_embossed_name=card.card_embossed_name.strip(),
        card_expiration_date=card.card_expiration_date,
        card_active_status=card.card_active_status,
        acct_curr_bal=f"{float(acct.acct_curr_bal):,.2f}" if acct else "",
        acct_credit_limit=f"{float(acct.acct_credit_limit):,.2f}" if acct else "",
        cust_name=(f"{cust.cust_first_name.strip()} {cust.cust_last_name.strip()}" if cust else ""),
        cust_id=cust.cust_id if cust else 0,
    )
    result.message = "Card record found"

File: pipeline/reimpl/java/runner/src/main/java/com/modernization/masquerade/runner/programs/Cocrdslc.java
package com.modernization.masquerade.runner.programs;

import com.modernization.masquerade.runner.ProgramRunner;

import java.util.*;

/**
 * Java reimplementation of COCRDSLC — CardDemo Credit Card View/Select Screen.
 *
 * <p>Mirrors {@code pipeline/reimpl/cocrdslc.py}. Python is the source of truth.
 *
 * <p>Standalone business method: {@link #processCardView}. The {@link #runVector}
 * method is a thin adapter that builds scenario-specific parameters and delegates.
 */
public class Cocrdslc implements ProgramRunner {

    // ── Inner data types ─────────────────────────────────────────────────

    static class CardRecord {
        final String cardNum;
        final String acctId;
        final String cvv;
        final String embossedName;
        final String expirationDate;
        final String activeStatus;

        CardRecord(String cardNum, String acctId, String cvv,
                   String embossedName, String expirationDate, String activeStatus) {
            this.cardNum = cardNum;
            this.acctId = acctId;
            this.cvv = cvv;
            this.embossedName = embossedName;
            this.expirationDate = expirationDate;
            this.activeStatus = activeStatus;
        }
    }

    static class AccountRecord {
        final String acctId;
        final double currBal;
        final double creditLimit;

        AccountRecord(String acctId, double currBal, double creditLimit) {
            this.acctId = acctId;
            this.currBal = currBal;
            this.creditLimit = creditLimit;
        }
    }

    static class XrefRecord {
        final String cardNum;
        final String custId;
        final String acctId;

        XrefRecord(String cardNum, String custId, String acctId) {
            this.cardNum = cardNum;
            this.custId = custId;
            this.acctId = acctId;
        }
    }

    static class CustomerRecord {
        final String custId;
        final String firstName;
        final String lastName;

        CustomerRecord(String custId, String firstName, String lastName) {
            this.custId = custId;
            this.firstName = firstName;
            this.lastName = lastName;
        }
    }

    static class CardDetailData {
        String cardNum = "";
        String acctId = "";
        String embossedName = "";
        String custName = "";
        String acctCurrBal = "";
    }

    static class CardViewResult {
        CardDetailData cardData = null;
        boolean error = false;
        String message = "";
        String xctlProgram = "";
        boolean returnToPrev = false;
    }

    // ── Repositories ────────────────────────────────────────────────────

    static class CardRepo {
        private final Map<String, CardRecord> cards;
        CardRepo(Map<String, CardRecord> cards) { this.cards = cards; }
        CardRecord find(String cardNum) { return cards.get(cardNum.trim()); }
    }

    static class AccountRepo {
        private final Map<String, AccountRecord> accounts;
        AccountRepo(Map<String, AccountRecord> accounts) { this.accounts = accounts; }
        AccountRecord find(String acctId) { return accounts.get(acctId); }
    }

    static class XrefRepo {
        private final Map<String, XrefRecord> byCard;
        XrefRepo(Map<String, XrefRecord> byCard) { this.byCard = byCard; }
        XrefRecord findByCard(String cardNum) { return byCard.get(cardNum.trim()); }
    }

    static class CustomerRepo {
        private final Map<String, CustomerRecord> customers;
        CustomerRepo(Map<String, CustomerRecord> customers) { this.customers = customers; }
        CustomerRecord find(String custId) { return customers.get(custId); }
    }

    // ── Standalone business method ──────────────────────────────────────

    /**
     * Process card view screen — mirrors Python's process_card_view.
     *
     * @param eibcalen      CICS EIBCALEN (0 = first entry, no commarea)
     * @param eibaid        aid key pressed (ENTER, PF3, etc.)
     * @param pgmContext    commarea program context (0 = initial)
     * @param fromProgram   commarea cdemo_from_program
     * @param cardNumInput  card number typed by user
     * @param preloadedCardNum card number preloaded from previous screen
     * @param cardRepo      card repository
     * @param accountRepo   account repository
     * @param xrefRepo      xref repository
     * @param customerRepo  customer repository
     * @return CardViewResult with all output fields
     */
    CardViewResult processCardView(
            int eibcalen,
            String eibaid,
            int pgmContext,
            String fromProgram,
            String cardNumInput,
            String preloadedCardNum,
            CardRepo cardRepo,
            AccountRepo accountRepo,
            XrefRepo xrefRepo,
            CustomerRepo customerRepo
    ) {
        CardViewResult result = new CardViewResult();

        if (eibcalen == 0) {
            result.returnToPrev = true;
            result.xctlProgram = "COSGN00C";
            return result;
        }

        if (eibaid.equals("PF3")) {
            String back = (fromProgram != null && !fromProgram.isEmpty()) ? fromProgram : "COMEN01C";
            result.xctlProgram = back;
            result.returnToPrev = true;
            return result;
        }

        if (pgmContext == 0) {
            String searchNum = (preloadedCardNum != null && !preloadedCardNum.isEmpty())
                    ? preloadedCardNum : cardNumInput;
            if (searchNum != null && !searchNum.trim().isEmpty()) {
                lookupCard(searchNum, cardRepo, accountRepo, xrefRepo, customerRepo, result);
            }
            return result;
        }

        // Process ENTER with card number
        return processEnter(cardNumInput, cardRepo, accountRepo, xrefRepo, customerRepo, result);
    }

    private CardViewResult processEnter(
            String cardNumInput,
            CardRepo cardRepo, AccountRepo accountRepo,
            XrefRepo xrefRepo, CustomerRepo customerRepo,
            CardViewResult result
    ) {
        if (cardNumInput == null || cardNumInput.trim().isEmpty()) {
            result.error = true;
            result.message = "Card number not provided";
            return result;
        }
        lookupCard(cardNumInput, cardRepo, accountRepo, xrefRepo, customerRepo, result);
        return result;
    }

    private void lookupCard(
            String cardNum,
            CardRepo cardRepo, AccountRepo accountRepo,
            XrefRepo xrefRepo, CustomerRepo customerRepo,
            CardViewResult result
    ) {
        CardRecord card = cardRepo.find(cardNum);
        if (card == null) {
            result.error = true;
            result.message = "Card " + cardNum.trim() + " not found";
            return;
        }

        AccountRecord acct = accountRepo.find(card.acctId);
        XrefRecord xref = xrefRepo.findByCard(card.cardNum);
        CustomerRecord cust = null;
        if (xref != null) {
            cust = customerRepo.find(xref.custId);
        }

        CardDetailData data = new CardDetailData();
        data.cardNum = card.cardNum;
        data.acctId = card.acctId;
        data.embossedName = card.embossedName.trim();
        data.custName = (cust != null) ? (cust.firstName.trim() + " " + cust.lastName.trim()) : "";
        data.acctCurrBal = (acct != null) ? String.format("%,.2f", acct.currBal) : "";

        result.cardData = data;
        result.message = "Card record found";
    }

    // ── Seed data ───────────────────────────────────────────────────────

    private static final Map<String, CardRecord> SEED_CARDS = new LinkedHashMap<>();
    private static final Map<String, AccountRecord> SEED_ACCOUNTS = new LinkedHashMap<>();
    private static final Map<String, XrefRecord> SEED_XREFS = new LinkedHashMap<>();
    private static final Map<String, CustomerRecord> SEED_CUSTOMERS = new LinkedHashMap<>();

    static {
        SEED_CARDS.put("4111111111110001", new CardRecord("4111111111110001", "10000001", "123", "JANE USER", "2028-12-31", "Y"));
        SEED_CARDS.put("4111111111110002", new CardRecord("4111111111110002", "10000001", "456", "BOB SMITH", "2027-06-30", "Y"));

        SEED_ACCOUNTS.put("10000001", new AccountRecord("10000001", 1500.00, 5000.00));

        SEED_XREFS.put("4111111111110001", new XrefRecord("4111111111110001", "100001", "10000001"));
        SEED_XREFS.put("4111111111110002", new XrefRecord("4111111111110002", "100001", "10000001"));

        SEED_CUSTOMERS.put("100001", new CustomerRecord("100001", "Jane", "User"));
    }

    // ── Thin runVector adapter ──────────────────────────────────────────

    @Override
    public Map<String, String> runVector(Map<String, String> inputs) {
        String scenario = inputs.getOrDefault("SCENARIO", "FIRST_ENTRY");

        CardRepo cardRepo = new CardRepo(new LinkedHashMap<>(SEED_CARDS));
        AccountRepo accountRepo = new AccountRepo(new LinkedHashMap<>(SEED_ACCOUNTS));
        XrefRepo xrefRepo = new XrefRepo(new LinkedHashMap<>(SEED_XREFS));
        CustomerRepo customerRepo = new CustomerRepo(new LinkedHashMap<>(SEED_CUSTOMERS));

        CardViewResult result;

        switch (scenario) {
            case "FIRST_ENTRY":
                result = processCardView(100, "ENTER", 0, "COCRDLIC",
                        "", "4111111111110001",
                        cardRepo, accountRepo, xrefRepo, customerRepo);
                break;

            case "LOOKUP_CARD":
                result = processCardView(100, "ENTER", 1, "COCRDLIC",
                        "4111111111110002", "",
                        cardRepo, accountRepo, xrefRepo, customerRepo);
                break;

            case "CARD_NOT_FOUND":
                result = processCardView(100, "ENTER", 1, "COCRDLIC",
                        "9999999999999999", "",
                        cardRepo, accountRepo, xrefRepo, customerRepo);
                break;

            case "PF3_RETURN":
                result = processCardView(100, "PF3", 1, "COCRDLIC",
                        "", "",
                        cardRepo, accountRepo, xrefRepo, customerRepo);
                break;

            default:
                result = processCardView(100, "ENTER", 0, "COCRDLIC",
                        "", "4111111111110001",
                        cardRepo, accountRepo, xrefRepo, customerRepo);
                break;
        }

        String cardNumOut = "";
        String acctIdOut = "";
        String embossedNameOut = "";
        String custNameOut = "";
        String acctCurrBalOut = "";
        if (result.cardData != null) {
            cardNumOut = result.cardData.cardNum;
            acctIdOut = result.cardData.acctId;
            embossedNameOut = result.cardData.embossedName;
            custNameOut = result.cardData.custName;
            acctCurrBalOut = result.cardData.acctCurrBal;
        }

        Map<String, String> out = new LinkedHashMap<>();
        out.put("ERROR", result.error ? "Y" : "N");
        out.put("MESSAGE", result.message);
        out.put("XCTL_PROGRAM", result.xctlProgram);
        out.put("RETURN_TO_PREV", result.returnToPrev ? "Y" : "N");
        out.put("CARD_NUM", cardNumOut);
        out.put("ACCT_ID", acctIdOut);
        out.put("EMBOSSED_NAME", embossedNameOut);
        out.put("CUST_NAME", custNameOut);
        out.put("ACCT_CURR_BAL", acctCurrBalOut);
        return out;
    }
}
